### PR TITLE
Handling reveal at PlayCardACtion instead of Cards

### LINKED ISSUE #216: Murkens archive bug
doesn't play random card from archive when the opponents draw pile is at 0

### LINKED ISSUE #434: Wild Wormhole shows a redundant out of format game message [cosmetic]
> romano2 plays Wild Wormhole, gaining 1 Æmber

> romano2 uses Wild Wormhole to play the top card of their deck : Mars Needs Æmber

> Wild Wormhole reveals Mars Needs Æmber

The thirs message seems redundant and does not say "player does xxx". It starts with card name:

image

### LINKED ISSUE #1601: Murkens / Fidgit vs Alpha / Omega
now when you use one of these cards ability, and opponent has alpha card (like sucker punch), I cannot even choose archives, topdeck is played

Today also my opponent played murkens, and I have dust imp and unlocked gateway.
Logs stated: 
Opponent uses murkens to play unlocked gateway
then somehow was played dust imp from my archive
then my opponent finished all his stuff
and at the end of the turn gateway ability was resolved.
and gateway remains in my archive

### LINKED ISSUE #1610: Fidget doesn't properly state what it discards
Steps to Repro:
1) Player 2 gets 2+ cards in their archive.
2) Player 1 reaps with Fidget, and chooses to discard from the archive.

Notice that sometimes the name of the card in chat doesn't match the name of the card that was actually discarded. (When this happened to me, the 2nd card in the archive was actually discarded but the 1st card name was displayed).

### LINKED ISSUE #2070: Murkens pulling from archive bug
Hi there, I used Murkens to play an opponent's random archived card. It pulled a Lollop the Titanic and said in the sidebar that I played Lollop, but it never appeared on my board and remained in the opponent's archives.

### LINKED ISSUE #2104: Wormhole Technician fizzles to play Logos alpha but does not log any message
Wormhole Technician fizzles to play Logos alpha but does not log any message

opening it after testing #1578

### PR DESCRIPTION
**Pull request by romanogit**
Fixes #216 
Fixes #434
Fixes #1601 
Fixes #1610
Fixes #2070 
Fixes #2104
Sentry: 
Fixes KEYTEKI-1CA
Fixes #KEYTEKI-1CA
Fixes [KEYTEKI-1CA]

### CODE REVIEW

**Comment by romanogit** [comment]
This fix is backward compatible at the moment, as it introduces a variable to reveal card. I will cleanup the other playCards and will push on this PR.

The core of the engine refactor is done already, if @jeremylarner and @cryogen want to review

File: server/game/GameActions/PlayCardAction.js [hunk 1]
```diff
@@ -58,6 +64,19 @@ class PlayCardAction extends CardGameAction {
                 context.game.resolveAbility(actionContext);
             });
         });
+
+        playEvent.addChildEvent(
+            context.game.getEvent('unamedEvent', { card, context }, () => {
+                if (playActions.length === 0 && this.revealOnIllegalTarget) {
+                    const message =
+                        this.revealOnIllegalTargetMessage ||
+                        '{0} uses {1} to only reveal and not play {2}';
```
Comment (cryogen) [inline server/game/GameActions/PlayCardAction.js:73]: What scenario does this cover? the wording isn't right and I need to understand what it's doing to suggest different wording

Comment (SteVader) [reply server/game/GameActions/PlayCardAction.js:73]: Cards like Chaos Portal and Wormhole Technician I think.

Comment (cryogen) [reply server/game/GameActions/PlayCardAction.js:73]: Yeah I think I'll defer to what I said on the other PR that introduces something similar. A custom message is overkill. It being generic would be fine

Comment (romanogit) [reply server/game/GameActions/PlayCardAction.js]: Fixed, thanks!

**Pull request merged**